content/library/advanced-features/advanced-widget-behavior.md

```diff
@@ -14,3 +16,252 @@ Widgets are magical and often work how you want. But they can have surprising be
+
+</Collapse>
+
+## Anatomy of a widget
+
+There are four parts to keep in mind when using widgets:
+
+1. The frontend component as seen by the user.
+2. The backend value or value as seen through `st.session_state`.
+3. The key of the widget used to access its value via `st.session_state`.
+4. The return value given by the widget's function.
+
+### Widgets are session dependent
+
+Widget states are dependent on a particular session (browser connection). The actions of one user do not affect the widgets of any other user. Furthermore, if a user opens up multiple tabs to access an app, each tab will be a unique session. Changing a widget in one tab will not affect the same widget in another tab.
+
+### Widgets return simple Python data types
+
+The value of a widget as seen through `st.session_state` and returned by the widget function are of simple Python types. For example, `st.button` returns a boolean value and will have the same boolean value saved in `st.session_state` if using a key.
+
+### Keys help distinguish widgets and access their values
+
+Widget keys serve two purposes:
+
+1. Distinguishing two otherwise identical widgets.
+2. Creating a means to access and manipulate the widget's value through `st.session_state`.
+
+Whenever possible, Streamlit updates widgets incrementally on the frontend instead of rebuilding them with each rerun. This means Streamlit assigns an ID to each widget from the arguments passed to the widget function. A widget's ID is based on parameters such as label, min or max value, default value, placeholder text, help text, and key. The page where the widget appears also factors into a widget's ID. If you have two widgets of the same type with the same arguments on the same page, you will get a `DuplicateWidgetID` error. In this case, assign unique keys to the two widgets.
+
+#### Streamlit can't understand two identical widgets on the same page
+
+```python
+# This will cause a DuplicateWidgetID error.
+st.button("OK")
+st.button("OK")
+```
+
+#### Use keys to distinguish otherwise identical widgets
+
+```python
+st.button("OK", key="privacy")
+st.button("OK", key="terms")
+```
+
+## Order of operations
+
+When a user interacts with a widget, the order of logic is:
+
+1. Its value in `st.session_state` is updated.
+2. The callback function (if any) is executed.
+3. The page reruns with the widget function returning its new value.
+
+If the callback function writes anything to the screen, that content will appear above the rest of the page. A callback function runs as a _prefix_ to the script rerunning. Consequently, that means anything written via a callback function will disappear as soon as the user performs their next action. Other widgets should generally not be created within a callback function.
```

Comment:
the word _prefix_ here feels a bit odd, although I understood what you meant

Reply:
Yeah, I don't know if there's a better way to say it. We explain elsewhere that "The callback is executed and then the page reruns." The problem is that people then think the next page load is starting with the rerun and not the callback. I've been using the word "prefix" on the forums since it seems intuitive to me as a way to describe that the next screen rendering _begins_ with the callback (correct) as opposed the the callback happening and the screen rendering _after_ that (incorrect).

Reply:
sounds good. I think this is a great paragraph, I don't have an immediate drop-in word to replace.